# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:** Cristóbal Alcázar, Gianina Salomó

**Fecha límite de entrega 📆:** Martes 4 de Abril.

**Tiempo estimado de dedicación:** 




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 100 palabras** . 🙏
` ` 

**Pregunta 1 (0.2 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones según lo visto en clases. Debe citar la slide donde cree que está la respuesta.**




` ` 
**Desafíos debido a sus propiedades**
- _Discreto_: Las letras de una palabra no entregan mucha información sobre la relación entre ellas.
- _Composicional_: El significado de una oración va más allá del significado individual de las palabras (depende del orden).
- _Disperso_: Existen infinitas combinaciones de palabras.

**Desafíos asociados a los datos de entrenamiento**

- _Costos de etiquetado_: En ML supervisado. Esto puede significar un costo de trabajo humano y tiempo.
- _Variaciones del dominio_: Los patrones del lenguaje pueden cambiar de un corpus a otro dependiendo del tema. Además, el lenguaje evoluciona en el tiempo, por lo que un modelo entrenado puede quedar desactualizado.

` ` 
_Referencia: Presentación "Natural Language Processing
Introduction", diapositivas 16 y 22_.

**Pregunta 2 (0.4 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.


` ` 
ML clásico requiere de una construcción manual de características (considera lingüística) y en general trabaja con bolsas de palabras. En cambio DL construye las características durante el entrenamiento y trabaja con word embeddings, además de requerir mucho mayor volumen de datos y por ende recursos. 

**Ejemplo: Clasificación de correo spam / no spam**

Con ML clásico se puede entrenar un modelo de Multinomial Naive Bayes (utiliza frecuencia de palabras en cada clase), mientras que con DL se puede entrenar una red MLP directamente desde el input, aprendiendo las características (word embeddings), más una capa de salida de clasificación.
` ` 

_No se utilizó ChatGPT. Referencias: Presentación "Natural Language Processing Introduction", diapositiva 27_, y cursos de Aprendizaje de máquinas y de Deep Learning.

 **Pregunta 3 (0.4 puntos) Según las primeras clases, ¿Qué método clásico nos permite rankear las similitudes existentes entre documentos?, ¿Cómo son las representaciones que genera y problemas que podrían experimentar estas soluciones simples?** 

` ` 
Un método que permite esto es la similitud coseno. Esta usa representaciones vectoriales de las palabras, que se pueden obtener mediante el vector space model, o "bag of words"; Esto logra que documentos de distinto largo residan en el mismo espacio vectorial. Como problema, se pierde la estructura lingüística lo cual no es ideal en algunas tareas. Se debe además suavizar para que todos los términos tengan ocurrencia y evitar indefiniciones (Ej: Agregar 0.5 a cada frecuencia).

Un problema del rankeo por similitud coseno es que no considera relevancia de los documentos, cosa que sí hacen los buscadores web.
` ` 

_Referencia: Presentación "Natural Language Processing
Vector Space Model and Information Retrieval", diapositivas 12-16_.

**Pregunta 4 (0.4 puntos) Usted se encuentra realizando un modelo de clasificación de sentimientos con texto, su jefe le señala que debe eliminar las palabras mas comunes para obtener una mejor clasificación. ¿Qué palabras le señala que elimine su jefe?, ¿es acaso esto una buena idea?.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.

` ` 
El jefe sugiere eliminar palabras con frecuencia sobre x%. No es una buena idea, ya que dentro de estas palabras podría haber palabras comunes como "no", que puede ser indicativas de sentimiento (en este caso negativo). Por ejemplo, frases como "no me gusta" o "no me siento bien" quedarían como "gusta" y "siento bien", que podrían clasificarse erróneamente como positivas, siendo que en realidad son negativas. Además, si se trabaja con textos cortos pueden generarse textos que pierden el sentido original. Por esto es importante revisar el caso particular y no usar siempre técnicas estándar (ej: eliminar stopwords de nltk).
` ` 

_No se usó ChatGPT. Referencia: Presentación "Natural Language Processing Vector Space Model and Information Retrieval", diapositiva 4, y video de la clase_.


**Pregunta 5 (0.6 puntos): Acorde al paper [A Vector Space Model for automatic indexing](https://dl.acm.org/doi/pdf/10.1145/361219.361220) un documento, $D_i$, puede ser definido formalmente como una tupla de términos, $(d_{i1}, d_{i2}, ..., d_{in})$, donde $d_{ij}$ representa el paso del j-esimo término. En clase vieron algunas formas medir los pesos de estos términos. Mencione cuales fueron y sus ventajas y desventajas.** 

` ` 
**Booleano**
- Ventaja: Simple de construir.
- Desventaja: No considera la frecuencia de los términos; Puede dar el mismo ranking a dos documentos, siendo que uno de ellos mencione más veces los términos de la query.

**Term Frequency**
- Ventaja: Sí considera la frecuencia de los términos.
- Desventaja: No normaliza las frecuencias; Términos frecuentes en todos los documentos debiesen penalizarse.

**Tf-Idf**
- Ventaja: Asigna mayor peso a términos frecuentes en el documento e infrecuentes en el corpus.
- Desventaja: Igual que los anteriores, no considera la lingüística. Además, palabras presentes en un documento (y no en el corpus) no son necesariamente relevantes para su tema.
` ` 

_Referencia: Presentación "Natural Language Processing Vector Space Model and Information Retrieval", diapositivas 12-14_.

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Tokens*.
- El **orden de los resultados** pueden variar dependiendo de su máquina, pero los valores de los resultados son los mismos.



**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [ ]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [ ]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [ ]:
import codecs
import re
import numpy as np
import pandas as pd
from string import punctuation

En caso de desarrollar la tarea desde colab, utilizar el siguiente código para cargar los archivos desde el drive:

```python
try:
    from google.colab import drive

    drive.mount("/content/drive")
    path = 'path/to/marcianeke.txt'
except: 
    print('Ignorando conexión drive-colab')
```

In [ ]:
# Archivo descargado directamente desde repositorio de GitHub
!wget https://raw.githubusercontent.com/alcazar90/CC6205-NLP/main/data/marcianeke.txt

--2023-03-23 11:46:22--  https://raw.githubusercontent.com/alcazar90/CC6205-NLP/main/data/marcianeke.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2084 (2.0K) [text/plain]
Saving to: ‘marcianeke.txt’

marcianeke.txt      100%[===================>]   2.04K  --.-KB/s    in 0s      

2023-03-23 11:46:22 (42.8 MB/s) - ‘marcianeke.txt’ saved [2084/2084]



` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene la letra de una canción del marcianeke 👽. Utilice este texto para realizar su primera tokenización y ver qué tan bien funciona su función. 

Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [ ]:
text = codecs.open('marcianeke.txt', 'r', 'UTF-8').read()
print(text)

Brr
Marcianeke
Vamo' a estar con Pailita
Dimelo má
Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar (brr)
Tussi, keta quiere' mezclar
Dimelo má

Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar
Tussi, keta pura quiere' mezclar
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má

Esperame que ahora entro yo
Y lo que pide yo lo traje
No visto de traje
Puro corte calle, no de maquillaje
Pronto coronamos y nos vamo' de viaje
Tanto hit que hago que lo' culo bajen
Ella se va de shopping
Sale positivo si se hace el doping
Baila twerk con un poco de popping
Los fardos en el botín

Si quieren letra llamen pa' mi booking
Generando, sigo en la mía lowkey
Cooking en el estudio con tu woman
Tanto whisky, pisco que hasta lo' vecinos toman
Si se tiran pa' aca puede que la arena coman
Ja, en el chanteo titulado sin diploma
Di-dimelo má
Di-di

Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [ ]:
def get_tokens(text):
  """
    Recibe una cadena de texto, y retorna una lista de Python
    con los tokens del texto de entrada.
  """
   # Elimina retorno de carro y tabs
  text = text.replace("\r", " ").replace("\t", " ")

  # Elimina dígitos
  text = re.sub('\d+', '', text) 

  # Genera lista separando por símbolos y espacios
  reg = r'(' + "|".join([re.escape(i) for i in punctuation]) + r'| )'
  text = re.split(reg, text)

  # Elimina espacios al principio y final de los tokens
  text = [t.strip() for t in text if t.strip()]
  return text

  ### Fin del código ###

In [ ]:
tokens = get_tokens(text)
print(tokens)

['Brr', 'Marcianeke', 'Vamo', "'", 'a', 'estar', 'con', 'Pailita', 'Dimelo', 'má', 'Ando', 'en', 'busca', 'de', 'una', 'criminal', '(', 'ah', ',', 'ah', ')', 'Esa', 'que', 'el', 'gatillo', 'le', 'gusta', 'jalar', '(', 'rata', '-', 'ta', ')', 'Que', 'le', 'guste', 'flotar', 'y', 'fumar', '(', 'brr', ')', 'Tussi', ',', 'keta', 'quiere', "'", 'mezclar', 'Dimelo', 'má', 'Ando', 'en', 'busca', 'de', 'una', 'criminal', '(', 'ah', ',', 'ah', ')', 'Esa', 'que', 'el', 'gatillo', 'le', 'gusta', 'jalar', '(', 'rata', '-', 'ta', ')', 'Que', 'le', 'guste', 'flotar', 'y', 'fumar', 'Tussi', ',', 'keta', 'pura', 'quiere', "'", 'mezclar', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Esperame', 'que', 'ahora', 'entro', 'yo', 'Y', 'lo', 'que', 'pide', 'yo', 'lo', 'traje', 'No', 'visto', 'de', 'traje', 'Puro', 'corte', 'calle', ',', 'no', 'de', 'maquillaje', 'Pro

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

**Supuestos en los que basamos nuestra tokenización**:

- Se asume que un token no necesariamente es una palabra con significado (como "ra" de "rata-ta" o "di' de "di-dimelo").
- Se asume que la puntuación y otros símbolos constituyen tokens por sí mismos.
- Se asume que los tildes determinan palabras diferentes y por ende tokens diferentes, por lo que se conservan (Ej: "papa" $\neq$ "papá").
- Se asume que, para el caso de las letras de canciones, los números no corresponden a "palabras" (o partes de ella), por lo que no se consideran como tokens. En otros casos, como detección de spam, se podría reemplazar los números por un token especial (ej: "dinero").
- Se asume que no existen palabras compuestas; Cada token corresponde a texto, cumpliendo las condiciones anteriores, separado por espacio. Por tanto, cada token corresponde a un unigrama. Esta decisión se tomó tomando en cuenta también el ejemplo de referencia.


**Comparación con `nltk`**

In [ ]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
print(nltk_tokens)
print(f"\n# Tokens propios: {len(tokens)}\n# Tokens nltk: {len(nltk_tokens)}")

assert tokens == nltk_tokens

['Brr', 'Marcianeke', 'Vamo', "'", 'a', 'estar', 'con', 'Pailita', 'Dimelo', 'má', 'Ando', 'en', 'busca', 'de', 'una', 'criminal', '(', 'ah', ',', 'ah', ')', 'Esa', 'que', 'el', 'gatillo', 'le', 'gusta', 'jalar', '(', 'rata', '-', 'ta', ')', 'Que', 'le', 'guste', 'flotar', 'y', 'fumar', '(', 'brr', ')', 'Tussi', ',', 'keta', 'quiere', "'", 'mezclar', 'Dimelo', 'má', 'Ando', 'en', 'busca', 'de', 'una', 'criminal', '(', 'ah', ',', 'ah', ')', 'Esa', 'que', 'el', 'gatillo', 'le', 'gusta', 'jalar', '(', 'rata', '-', 'ta', ')', 'Que', 'le', 'guste', 'flotar', 'y', 'fumar', 'Tussi', ',', 'keta', 'pura', 'quiere', "'", 'mezclar', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Di', '-', 'dimelo', 'má', 'Esperame', 'que', 'ahora', 'entro', 'yo', 'Y', 'lo', 'que', 'pide', 'yo', 'lo', 'traje', 'No', 'visto', 'de', 'traje', 'Puro', 'corte', 'calle', ',', 'no', 'de', 'maquillaje', 'Pro

- Los resultados generados con el tokenizador propio son iguales a los generados por `nltk`, al menos para este caso.

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [ ]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los tokens de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [ ]:
def get_vocab(dataset):
  """
    Retorna una lista con el vocabulario de una colección de
    documentos. Se utiliza la función get_tokens() para extraer
    el vocabulario de los documentos.
    
    Nota: Utilice 'np.concatenate' porque estaba dentro
    de las librerías que se importaron al principio.
  """
  ### Aquí inicia tu código ###
  vocab = set([token for doc in dataset for token in get_tokens(doc)])
  return list(vocab)
  ### Aquí termina tu código ###

***Test:***

In [ ]:
vocab = get_vocab(dataset)
print(vocab)

['like', 'human', 'Spanish', 'my', 'is', 'language', 'languages', 'programming', 'I', 'favorite']


``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

In [ ]:
esperado = [
    'favorite', 'Spanish', 'language', 'I', 'like', 'programming', 'languages',
    'my', 'human', 'is'
]

assert sorted(esperado) == sorted(vocab)

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

**Explique sus reglas aquí**:

Stopwords
- Una forma de armar una lista de stopwords es tomando las "x" primeras palabras del ranking de palabras más usadas en inglés. 
- Para este ejercicio, tomamos como referencia el sitio [word frequency](https://www.wordfrequency.info/), el cual tiene disponible en forma gratuita las primeras 5.000 palabras del ranking de palabras en inglés.
- En nuestra implementación, se utilizó el parámetro "top" (por defecto 20, valor utilizado en la implementación) para modificar la cantidad de palabras frecuentes a retornar.
- Se debe señalar además que para un caso particular, en lugar de usar listas pre elaboradas de stop words, o usar las palabras más frecuentes de todo el idioma (como hicimos en este caso), se puede usar las palabras más frecuentes dentro del corpus de entrenamiento. Por ejemplo, en casos de clasificación (como sentimientos), se puede considerar como stop words aquellas palabras que tengan alta frecuencia en las distintas clases (ya que no facilitan la discriminación entre una clase y otra). Sin embargo, para este ejercicio decidimos usar las palabras más frecuentes del idioma inglés, y no las más frecuentes de este caso específico, ya que el corpus entregado es muy pequeño.

Stemming
- Para este ejercicio, se considera las reglas del algoritmo de Porter vistas en clase (Presentación "Natural Language Processing Vector Space Model and Information Retrieval", diapositiva 5).
- Se debe considerar que el [algoritmo de Porter](https://nlp.stanford.edu/IR-book/html/htmledition/stemming-and-lemmatization-1.html) es mucho más complejo, y considera muchas más reglas y heurísticas, por lo que la solución implementada puede entregar reemplazos erróneos (Ej: "is" $\rightarrow$ "i").
- De todas formas, la solución implementada es escalable, y se le puede agregar más reglas o un llamado a otra función de procesamiento.
- Finalmente, destacar que el algoritmo de Porter se encuentra implementado en `nltk`, pero de todas formas muchas veces no suele ser la solución más óptima al momento de preprocesar texto y generar tipos; Otra alternativa es el uso de _lemmatización_ (también disponible en `nltk`, que se basa en reglas y tablas de búsqueda, y también lleva las palabras a sus raíces.

In [ ]:
def get_most_frequent_words(top=20):
  """
  Retorna una lista de Python con las "top" palabras frecuentes
  en inglés, según el sitio wordfrequency
  """
  df = pd.read_excel("https://www.wordfrequency.info/samples/wordFrequency.xlsx",
                     sheet_name="1 lemmas")
  
  return df["lemma"].head(top).to_list()

# Test
stopwords = get_most_frequent_words()
print(stopwords)

['the', 'be', 'and', 'a', 'of', 'to', 'in', 'i', 'you', 'it', 'have', 'to', 'that', 'for', 'do', 'he', 'with', 'on', 'this', "n't"]


In [ ]:
def pre_processing(vocab):
  """
    Recibe un vocabulario (lista de tokens únicos), y retorna
    un vocabulario procesado, donde se eliminan stopwords
    y se aplica una parte del algoritmo de Porter de Stemming
  """

  ### Aquí inicia tu código ###

  # Remover stopwords
  stopwords = get_most_frequent_words()
  res = [w for w in vocab if w.lower() not in stopwords]

  # Se aplica parte del algoritmo de Porter
  # Esto podría expandirse a más reglas según se requiera
  res = pd.Series(res).replace(
      [r"sses+$|SSES+$", r"ies+$|IES+$", r"s+$|S+$"],
      ["ss", "i", ""],
      regex=True,
  ).unique()
  
  return list(res)
  ### Aquí termina tu código ###
    

In [ ]:
vocab = pre_processing(vocab)
print(vocab)

['like', 'human', 'Spanish', 'my', 'i', 'language', 'programming', 'favorite']



**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
 


Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [ ]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [ ]:
def bag_of_words(dataset):
  """
    Retorna matriz de bag-of-words o matriz de ocurrencia por documento.

    Nota: falta incluir el preprocessing dentro del vocabulario y la
    obtención de tokens.
    
  """
  ### Aquí inicia tu código ###
  # Obtener vocabulario del dataset
  vocab = get_vocab(dataset)

  # Crear matriz de ocurrencia: número de documentos x largoo vocabulario
  bow = np.zeros((len(dataset), len(vocab)))

  # Iterar por cada documento y poblar matriz de ocurrencia
  row_names = []

  for idx, doc in enumerate(dataset):

    row_names.append('d' + str(idx))
    # obtener tokens por documento según nuestra función get_tokens()
    tokens_in_doc = get_tokens(doc)

    for token in tokens_in_doc:
      # según idx del documento y posición del token dentro del vocabulario
      # las columnas siguen el orden de la lista de vocabulario
      bow[idx, vocab.index(token)] += 1

  # Retornar como pandas dataframe
  return pd.DataFrame(bow, columns=vocab, index=row_names).astype(int)
  ### Aquí termina tu código ###

***Test:***

In [ ]:
dataset_bow = bag_of_words(dataset)
dataset_bow

maullar  perro  y  después  duerme  El  gato  la  empieza  se  despierta  \
d0        0      1  1        1       1   1     0   1        0   2          0   
d1        0      1  1        1       0   1     0   0        1   1          1   
d2        0      1  1        1       0   1     0   1        0   1          0   
d3        0      0  1        1       1   1     1   1        0   2          0   
d4        1      0  1        1       0   1     1   0        1   1          1   
d5        0      0  1        1       0   1     1   1        0   1          0   

    ladra  a  maulla  comida  ladrar  come  
d0      0  0       0       1       0     1  
d1      0  1       0       0       1     0  
d2      1  0       0       1       0     1  
d3      0  0       0       1       0     1  
d4      0  1       0       0       0     0  
d5      0  0       1       1       0     1

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [ ]:
def calc_tf(dataset_bow):
  """
  Retorna matriz de frecuencia por término en formato pd.DataFrame.

  Esto es simplemente normalizar por fila las ocurrencias. Luego,
  volver a normalizar por máxima frecunecia."""
  ### Aquí inicia tu código ###

  # Normalizar por fila para obtener la frecuencias t_{ij}
  # t_{ij} = frecuencia del término "i" en el documento "j"
  tf = dataset_bow.div(dataset_bow.sum(axis=1), axis=0)

  # Normalizar por el término con máxima frecuencia en el documento
  # max_{i}(tf_{ij}) = mayor frecuencia en documento "j"
  return tf.div(tf.max(axis=1), axis=0)

  ### Aquí termina tu código ###

***Test:***

In [ ]:
tf = calc_tf(dataset_bow)
tf

maullar  perro    y  después  duerme   El  gato   la  empieza   se  \
d0      0.0    0.5  0.5      0.5     0.5  0.5   0.0  0.5      0.0  1.0   
d1      0.0    1.0  1.0      1.0     0.0  1.0   0.0  0.0      1.0  1.0   
d2      0.0    1.0  1.0      1.0     0.0  1.0   0.0  1.0      0.0  1.0   
d3      0.0    0.0  0.5      0.5     0.5  0.5   0.5  0.5      0.0  1.0   
d4      1.0    0.0  1.0      1.0     0.0  1.0   1.0  0.0      1.0  1.0   
d5      0.0    0.0  1.0      1.0     0.0  1.0   1.0  1.0      0.0  1.0   

    despierta  ladra    a  maulla  comida  ladrar  come  
d0        0.0    0.0  0.0     0.0     0.5     0.0   0.5  
d1        1.0    0.0  1.0     0.0     0.0     1.0   0.0  
d2        0.0    1.0  0.0     0.0     1.0     0.0   1.0  
d3        0.0    0.0  0.0     0.0     0.5     0.0   0.5  
d4        1.0    0.0  1.0     0.0     0.0     0.0   0.0  
d5        0.0    0.0  0.0     1.0     1.0     0.0   1.0

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [ ]:
def calc_idf(dataset_bow):
  """
  Retorna diccionario con los score idf por cada token en el
  vocabulario en dataset_bow.

    - dataset_bow [pd.DataFrame]: bag-of-words dataframe, o matriz
    de ocurrencia, computado por la función bag_of_words()
  """
  ### Aquí inicia tu código ###

  # Extraemos el vocabulario de las clumnas
  vocab = dataset_bow.columns

  # Matriz de ocurrencia (C) y número de documentos (N)
  C = dataset_bow.to_numpy()
  N, _ = C.shape

  # Colapsamos las filas en un vector ni y computamos idf
  ni = np.sum(np.where(C > 0.0, 1.0, 0.0), axis=0)
  idf = np.log10(N/ ni)

  # Retornamos idf por vocabulario en un diccionario
  return {vocab[idx]: val for idx, val in enumerate(idf)}

  ### Aquí termina tu código ###

***Test:***

In [ ]:
idf = calc_idf(dataset_bow)
idf

{'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'y': 0.0,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'El': 0.0,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'empieza': 0.47712125471966244,
 'se': 0.0,
 'despierta': 0.47712125471966244,
 'ladra': 0.7781512503836436,
 'a': 0.47712125471966244,
 'maulla': 0.7781512503836436,
 'comida': 0.17609125905568124,
 'ladrar': 0.7781512503836436,
 'come': 0.17609125905568124}

***Resultado esperado***: 

```python
{'El': 0.0, 
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [ ]:
def calc_tf_idf(tf, idf):
  """
  Retorna matriz de scores tf-idf.

    tf [pd.DataFrame]: term frequency matrix en formato dataframe.
    idf [dict]: diccionario (vocab, idf).

    Nota: vocab concuerda con los nombres de las columnas de tf.
  """
  ### Aquí inicia tu código ###

  # Dado que la llave de idf == tf.columns:
  return tf.mul(idf)
  ### Aquí termina tu código ### 

***Test.***

In [ ]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

maullar     perro    y  después    duerme   El      gato        la  \
d0  0.000000  0.150515  0.0      0.0  0.238561  0.0  0.000000  0.088046   
d1  0.000000  0.301030  0.0      0.0  0.000000  0.0  0.000000  0.000000   
d2  0.000000  0.301030  0.0      0.0  0.000000  0.0  0.000000  0.176091   
d3  0.000000  0.000000  0.0      0.0  0.238561  0.0  0.150515  0.088046   
d4  0.778151  0.000000  0.0      0.0  0.000000  0.0  0.301030  0.000000   
d5  0.000000  0.000000  0.0      0.0  0.000000  0.0  0.301030  0.176091   

     empieza   se  despierta     ladra         a    maulla    comida  \
d0  0.000000  0.0   0.000000  0.000000  0.000000  0.000000  0.088046   
d1  0.477121  0.0   0.477121  0.000000  0.477121  0.000000  0.000000   
d2  0.000000  0.0   0.000000  0.778151  0.000000  0.000000  0.176091   
d3  0.000000  0.0   0.000000  0.000000  0.000000  0.000000  0.088046   
d4  0.477121  0.0   0.477121  0.000000  0.477121  0.000000  0.000000   
d5  0.000000  0.0   0.000000  0.000000  0.000000  0.778151  0.176091   

      ladrar      come  
d0  0.000000  0.088046  
d1  0.778151  0.000000  
d2  0.000000  0.176091  
d3  0.000000  0.088046  
d4  0.000000  0.000000  
d5  0.000000  0.176091

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [ ]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###

  # v1.dot(v1) ** 0.5 <-> np.sqrt((v1 ** 2).sum())
  return v1.dot(v2) / (v1.dot(v1)**.5 * v2.dot(v2)**.5)
  
  ### Aquí termina tu código ### 

# Número de documentos en le dataset 
N_DOC = len(dataset)

# Crear matriz N_DOC x N_DOC para guardar similitudes entre documentos
similarity_matrix = np.zeros((N_DOC, N_DOC))  

for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity


In [ ]:
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


Primero, dos _sanity checks_ de que la matriz de documentos fue computada correctamente.

- [X] La matriz es simetrica dado que `cos(d1,d2)=cos(d2,d1)`.
- [X] La diagonal de la matriz es 1 (i.e. `cos(d1,d1)=1`).


Dado que es simetrica nos podemeos fijar en la parte triangular superior
de la matriz,

In [ ]:
print(np.triu(similarity_matrix))

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.         1.         0.08686457 0.         0.4952126  0.        ]
 [0.         0.         1.         0.1632828  0.         0.11787732]
 [0.         0.         0.         1.         0.12032418 0.3223436 ]
 [0.         0.         0.         0.         1.         0.08686457]
 [0.         0.         0.         0.         0.         1.        ]]


Según la similitud por coseno, y el proceso de tokenización realizado, los
documentos más similares son la tupla $(d_{0}$ y $d_{3})$. Reportando una
métrica de $0.77967$.

Observemos estos documentos:

In [ ]:
# Los dos más similares
print(d0)
print(d3)

El perro se come la comida y después se duerme
El gato se come la comida y después se duerme


In [ ]:
# Los siguientes dos más similares
print(d1)
print(d4)

El perro se despierta y después empieza a ladrar
El gato se despierta y después empieza a maullar


**Conclusiones**

- Se observa que los dos documentos con mayor similaridad entre sí son los documentos `d0` y `d3`, los cuales difieren en 1 sola palabra. Por tanto, sí hace sentido que la similaridad calculada sea la más alta. En este caso la similaridad es de 0.78.
- La segunda pareja de documentos más similares, son los documentos `d1` y `d4`, los cuales difieren en 2 palabras; Por tanto, hace sentido que tengan una similaridad menor que el caso anterior. En este caso la similaridad baja mucho, a 0.5, aunque solo se agregue 1 palabra de diferencia. Esto tiene sentido, ya que la cantidad de vocabulario es pequeña.

![gato](https://live.staticflickr.com/4652/38904147065_0b6c446945_b.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**